### DISCUSSION TITLE
`UnstructuredGrid` object has no attribute `compute_normals`

### DISCUSSION THREAD

**Discussion by ldv1** [pyvista/pyvista#4944] [Q&A]
Hi,

I read in a vtu unstructured file. It contains only triangles in 3D space that describe a surface mesh. It is correctly rendered by pyvista. That's fine. However, when I compute the normals:

```python
mesh = pv.read("myfile.vtu")
mesh.compute_normals(inplace=True)
```

and I got

`AttributeError: 'UnstructuredGrid' object has no attribute 'compute_normals'`

Is there a way around this issue?
Thanks for any help!

**Comment by MatthewFlamm** [comment]
Can you use `extract_surface` to get a `PolyData` first?

**Comment by ldv1** [comment]
Thanks for your answer.

However, if I use it that way
```python
mesh = pv.read("myfile.vtu")
mesh = mesh.extract_surface()
plotter.add_mesh(mesh)
mesh.compute_normals(inplace=True)
arrows = mesh["Normals"]
plotter.add_arrows(mesh.points, arrows)
```

I got

`ValueError: center and direction arrays must have the same shape`

triggered by `plotter.add_arrows(mesh.points, arrows)
`

**Reply by tkoyama010** [reply]
In my mind, you think you are getting the normals of the POINT, but the shape does not match the point because you are getting the normals of the CELL.
You can clearly get the normals of the point by using `point_normals`.
```python
import numpy as np
import pyvista as pv
from pyvista import examples

plotter = pv.Plotter()
mesh = examples.cells.Polygon()
plotter.add_mesh(mesh)
mesh = mesh.extract_surface()
arrows = mesh.point_normals
plotter.add_arrows(mesh.points, arrows)
plotter.show()
```
![image](https://github.com/pyvista/pyvista/assets/7513610/72f34025-d282-4138-8351-080561ed93e3)

**Reply by ldv1** [reply]
You are right. Now it works. Many Thanks!
It is nevertheless unclear (undocumented?) why the normals are computed when we call `extract_surface`.

With `mesh.compute_normals`, I can specify `consistent_normals=False` if I know that the points of the triangles are consistently ordered. I am mentioning this because the normals in my case do not seem to be consistent after my call to `extract_surface`. Or is there a way to avoid pyvista to enforce what it consider to be a consistent ordering in this case?

**Reply by MatthewFlamm** [reply]
PyVista is using vtkDataSetSurfaceFilter under the hood. PyVista does not explicitly do any normal computation outside of this.  So this must happen inside the vtk class.
  See https://vtk.org/doc/nightly/html/classvtkDataSetRegionSurfaceFilter.html  and

https://github.com/pyvista/pyvista/blob/release/0.42/pyvista/core/filters/data_set.py#L4675-L4761

I think you can call `compute_normals` afterwards to get more control of the normals calculation.
